debezium-core/src/main/java/io/debezium/pipeline/signal/actions/SignalActionProvider.java

```diff
@@ -0,0 +1,38 @@
+/*
+ * Copyright Debezium Authors.
+ *
+ * Licensed under the Apache Software License version 2.0, available at http://www.apache.org/licenses/LICENSE-2.0
+ */
+package io.debezium.pipeline.signal.actions;
+
+import io.debezium.config.CommonConnectorConfig;
+import io.debezium.pipeline.EventDispatcher;
+import io.debezium.pipeline.spi.Partition;
+import io.debezium.spi.schema.DataCollectionId;
+
+/**
+ * This interface is used to provide custom signal actions:
+ * Implementations must:
+ * define the name of the action in {@link #getName()},
+ * provide the implementation of the action in the {@link #createAction(EventDispatcher, CommonConnectorConfig)} method.
+ *
+ * @author Mario Fiore Vitale
+ */
+public interface SignalActionProvider {
```

Comment:
@mfvitale Why do we need provider classes? Can't implementations of class `Action` be service loaded and the `getName()` method added to them? I don't see an advatage of using indirect apprach in this case.

Reply:
OTOH, yes let's make it consistent with other providers we use. Still do we need per-action provider? How about having a provider that will look like

```java
interface SignalActionProvider {

    // returns map of name of signal, action
    Map<String, <P extends Partition> SignalAction<P>> createActions(EventDispatcher<P, ? extends DataCollectionId> dispatcher, CommonConnectorConfig connectorConfig);
}
```

Reply:
The actions requires the dispatcher instance, so in this case direct instantiation with ServiceLoader is not possible. 

> @mfvitale Why do we need provider classes? Can't implementations of class `Action` be service loaded and the `getName()` method added to them? I don't see an advatage of using indirect apprach in this case.

The actions requires the dispatcher instance, so in this case direct instantiation with ServiceLoader is not possible.

Reply:
> OTOH, yes let's make it consistent with other providers we use. Still do we need per-action provider? How about having a provider that will look like
> 
> ```java
> interface SignalActionProvider {
> 
>     // returns map of name of signal, action
>     Map<String, <P extends Partition> SignalAction<P>> createActions(EventDispatcher<P, ? extends DataCollectionId> dispatcher, CommonConnectorConfig connectorConfig);
> }
> ```

I think it can be a good alternative even in case of user extension seem a good interface.

Reply:
@mfvitale Yes, please do.

Reply:
> The actions requires the dispatcher instance, so in this case direct instantiation with ServiceLoader is not possible.
> 
> > @mfvitale Why do we need provider classes? Can't implementations of class `Action` be service loaded and the `getName()` method added to them? I don't see an advatage of using indirect apprach in this case.
> 
> The actions requires the dispatcher instance, so in this case direct instantiation with ServiceLoader is not possible.

This would be solvable by executing late initializtion - in the same way as we use `configure` methods. Still it is unnecessary as the other option I talked about seems like best of both worlds.